pkg/controller/storagecluster/external_resources.go

```diff
@@ -54,39 +54,38 @@ func (r *ReconcileStorageCluster) ensureExternalStorageClusterResources(instance
 	if err != nil {
 		return err
 	}
-	ownerRef := metav1.OwnerReference{
-		UID:        instance.UID,
-		APIVersion: instance.APIVersion,
-		Kind:       instance.Kind,
-		Name:       instance.Name,
-	}
 	var data []ExternalResource
 	err = json.Unmarshal(found.Data[externalClusterDetailsKey], &data)
 	if err != nil {
 		reqLogger.Error(err, "could not parse json blob")
 		return err
 	}
-	scs, err := r.newStorageClasses(instance)
-	if err != nil {
-		reqLogger.Error(err, "failed to create StorageClasses")
-		return err
-	}
-	err = r.createExternalStorageClusterResources(ownerRef, scs, data, instance, reqLogger)
-	if err != nil {
-                reqLogger.Error(err, "could not create ExternalStorageClusterResource")
-                return err
-        }
-	// creating all the storageClasses once we set the values
-	err = r.createStorageClasses(scs, reqLogger)
+	err = r.createExternalStorageClusterResources(data, instance, reqLogger)
 	if err != nil {
+		reqLogger.Error(err, "could not create ExternalStorageClusterResource")
 		return err
 	}
 	instance.Status.ExternalSecretFound = true
 	return nil
 }
 
-// createExternalStorageClusterResources creates external cluster resources
-func (r *ReconcileStorageCluster) createExternalStorageClusterResources(ownerRef metav1.OwnerReference, scs []*storagev1.StorageClass, data []ExternalResource, instance *ocsv1.StorageCluster, reqLogger logr.Logger) error {
+// createExternalStorageClusterResources create the needed external cluster resources
+func (r *ReconcileStorageCluster) createExternalStorageClusterResources(
+	data []ExternalResource, instance *ocsv1.StorageCluster, reqLogger logr.Logger) error {
+	ownerRef := metav1.OwnerReference{
+		UID:        instance.UID,
+		APIVersion: instance.APIVersion,
+		Kind:       instance.Kind,
+		Name:       instance.Name,
+	}
+	scs, err := r.newStorageClasses(instance)
+	if err != nil {
+		reqLogger.Error(err, "failed to create StorageClasses")
+		return err
+	}
+	// make a ZERO length array with capacity same as 'scs' length
+	// will store only the provided StorageClasses
+	availableSCs := make([]*storagev1.StorageClass, 0, len(scs))
```

Comment:
Why not making the length dynamic instead of calling `newStorageClasses()`? It seems overkill to call it just to get a fixed length...

Reply:
The above `make([]<type>, 0, size)` call will create a ZERO length array, with capacity of `size`. It is dynamic (as long as you call `append()` function to add items into the array).
The only advantage of the above `make` is while doing an `append`, the array don't have to relocate (no extra garbage collection required) itself in order to accommodate newer elements.
More details here: https://stackoverflow.com/questions/33834742/remove-and-adding-elements-to-array-in-go-lang

Reply:
I understand the underlying effect, I'm just not sure if it's worth the optimization given that we call `r.newStorageClasses(instance)`. I'd say that's almost confusing too unless you add a comment explaining the rationale.

Reply:
About calling `newStorageClasses()` function, we require the actual StorageClasses to be created first and then we use each SC to fix their parameters according to the JSON input. We can see those changes in [Line#147](https://github.com/openshift/ocs-operator/pull/701/files#diff-e5a5352dff74ec9aa443933a8a24fd0eR147)

Reply:
I agree, this looks like a needless optimization.

Reply:
Changed...